### ISSUE TITLE
[Bug][Many Classes] Concurrent batch import load leads to class creation timeouts

### ISSUE THREAD

**Issue by etiennedi** [weaviate/weaviate#2835]
## tl;dr

- `db.IndexLock` protects against race conditions when adding/removing classes 
- There seem to be places that don't just hold the `IndexLock` for the purpose above, but hold it while performing expensive operations
- Since IndexLock is an `RWMutex` frequent class updates while running batch imports leads to timeouts

## Long explanation

We obtained a trace (`go tool pprof trace`) of a running class create operation that timed out. The last logged unit of work (before going into 45s of synchronization block) was this line: https://github.com/weaviate/weaviate/blob/master/adapters/repos/db/migrator.go#L86

The following code block is an excellent example of what the `indexLock` is meant to achieve. It should prevent race conditions when reading from or writing into the hash map that holds indexes by name: https://github.com/weaviate/weaviate/blob/ce78f0add6fca90786372540ee3d27683f2bcd05/adapters/repos/db/migrator.go#L86-L89

As a result, this lock should only ever be held for a few nanoseconds or microseconds at the most.

However, we have identified at least two other places where the lock can be held for a considerable time. For example here the lock is held during an entire batch operation which can easily take multiple seconds:

https://github.com/weaviate/weaviate/blob/ce78f0add6fca90786372540ee3d27683f2bcd05/adapters/repos/db/batch.go#L33-L34

The same is true for reference batches: 

https://github.com/weaviate/weaviate/blob/ce78f0add6fca90786372540ee3d27683f2bcd05/adapters/repos/db/batch.go#L71-L72

## Reproduction

The following new chaos pipeline [can reproduce the error](https://github.com/weaviate/weaviate-chaos-engineering/actions/runs/4544325624/jobs/8010196767). The pipeline is built in a way that an expensive batch import (made expensive by choosing high dimensionality and large batches) runs in the background. In the foreground, we're adding and deleting classes. The pipeline fails as soon as a specific threshold (currently set to 5s) is exceeded for any operation.

## Possible Solution

The two examples mentioned above hold the lock because what follows is an iteration over the map. However, the lock is not intended to guarantee that indexes don't change. It's meant to prevent a race condition on the map. So an equally good solution could be to copy the pointers from the map into a local structure while holding the lock, then releasing the lock and doing the expensive work without holding the lock.

**Comment by donomii** [comment]
I've used a lock free map library for situations like this, it might work for weaviate: https://github.com/cornelk/hashmap.

It's faster on reads, slower on writes than the usual go map

**Comment by dudanogueira** [comment]
Apparently, [this was fixed already](https://deepwiki.com/search/tldr-dbindexlock-protects-agai_5fdfaa2c-a04e-421a-ae63-0b7db0248b4f?mode=fast) 🤔
